### PR TITLE
[Fix pull-local-node-e2e] Install CNI and Enable containerd for kubekins docker

### PR DESCRIPTION
**Pull request by adisky**
Signed-off-by: Aditi Sharma <adi.sky17@gmail.com>
/sig node
/sig testing
/cc @aojea @dims 

#### What type of PR is this?
/kind failing-test

#### What this PR does / why we need it:
Add support to install CNI when running in CI, also enables cri for docker so Kubelet can connect to docker installed containerd

#### Which issue(s) this PR fixes:

Ref https://github.com/kubernetes/test-infra/issues/24715

#### Special notes for your reviewer:
This PR alone is not sufficient to fix job, we need to set some env variables 
test-infra PR :- https://github.com/kubernetes/test-infra/pull/25352

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: hack/local-up-cluster.sh [hunk 1]
```diff
@@ -1044,6 +1044,93 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
...
+  cd /workspace 
+  TMPROOT=$(mktemp -d)
+  git clone https://github.com/containernetworking/plugins.git "${TMPROOT}"/plugins
+  pushd "${TMPROOT}"/plugins
+  git checkout v1.0.1
+  ./build_linux.sh
+  mkdir -p $CNI_DIR
+  cp -r ./bin $CNI_DIR
+  mkdir -p $CNI_CONFIG_DIR
+  cat << EOF | tee $CNI_CONFIG_DIR/10-containerd-net.conflist
```
Comment (aojea) [inline hack/local-up-cluster.sh:1075]: we use this in kind to download the binaries and use just the ones we need

https://github.com/kubernetes-sigs/kind/blob/80a6a355a2065823bbdae5048abf76a4f2a3ec9d/images/base/Dockerfile#L171-L188

int his case, you just need to replace `ptp` by `bridge` in the script

File: hack/local-up-cluster.sh [hunk 2]
```diff
@@ -1044,6 +1044,93 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
...
+cd $PDIR
+echo "Enabling IPV6 for Docker."
+# configure the daemon with ipv6
+mkdir -p /etc/docker/
+    cat <<EOF >/etc/docker/daemon.json
+{
+  "ipv6": true,
+  "fixed-cidr-v6": "fc00:db8:1::/64"
+}
+EOF
```
Comment (aojea) [inline hack/local-up-cluster.sh:1121]: why is docker required? is not about containerd only?

Comment (adisky) [reply hack/local-up-cluster.sh:1121]: we need docker to run kubetest

File: hack/local-up-cluster.sh (hunk 1 above)
Comment (adisky) [reply hack/local-up-cluster.sh:1075]: Thanks, this is really rough script 🙈 , will refine later once it starts passing

**Comment by adisky** [comment]
@aojea can you take a look now?

File: hack/local-up-cluster.sh [hunk 3]
```diff
@@ -1044,6 +1112,20 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
 
   # kubekins has a special directory for docker root
   DOCKER_ROOT="/docker-graph"
+
+  # to use docker installed containerd as kubelet container runtime
+  # we need to enable cri and install cni
+  # install cni for docker in docker
+  # containerd 1.4.12 installed by docker supports CNI version 0.4.0
+  CNI_VERSION=0.4.0
+  install_cni "${CNI_VERSION}"
```
Comment (aojea) [inline hack/local-up-cluster.sh:1121]: this should be optional to avoid conflicts with current installations, right?

Comment (aojea) [reply hack/local-up-cluster.sh:1121]: or just check that `CNI_CONFIG_DIR` is empty

File: hack/local-up-cluster.sh [hunk 4]
```diff
@@ -1044,6 +1112,20 @@ if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then
 
   # kubekins has a special directory for docker root
   DOCKER_ROOT="/docker-graph"
+
+  # to use docker installed containerd as kubelet container runtime
+  # we need to enable cri and install cni
+  # install cni for docker in docker
+  # containerd 1.4.12 installed by docker supports CNI version 0.4.0
+  CNI_VERSION=0.4.0
```
Comment (aojea) [inline hack/local-up-cluster.sh:1120]: this variable is duplicate and is not really needed, just hardcode 0.4.0 to the config file

>   "cniVersion": "${CNI_VERSION}",
to
>   "cniVersion": "0.4.0",

File: hack/local-up-cluster.sh (hunk 3 above)
Comment (aojea) [reply hack/local-up-cluster.sh:1121]: yeah, definitively I don't think is good idea downloading things by default,

Comment (aojea) [reply hack/local-up-cluster.sh:1121]: nevermind, this is correct, it already in 
`if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]]; then`

Comment (adisky) [reply hack/local-up-cluster.sh:1121]: this is only done for 
```
if [[ "${KUBETEST_IN_DOCKER:-}" == "true" ]];
```
which runs in the CI where containerd or any other cri not installed, so won't conflict with any current installtion

File: hack/local-up-cluster.sh (hunk 4 above)
Comment (adisky) [reply hack/local-up-cluster.sh:1120]: actually there is a reason not directly hardcoded to config file, the function `install_cni` is intended to be used by directly by the user of the script, so they can set cni version as per there requirement

I would have hardcoded it to 1.0.0 but the containerd installed by docker in kubekins image does't support it

Comment (aojea) [reply hack/local-up-cluster.sh:1120]: this version is only the one that goes in the manifest conflist, is not needed

Comment (aojea) [reply hack/local-up-cluster.sh:1120]: you have CNI_VERSION at the beginning to represent the version of the binary

Comment (adisky) [reply hack/local-up-cluster.sh:1120]: but what i found is `CNI_VERSION` & `CNI_PLUGINS_VERSION` both are different 
`CNI_PLUGINS_VERSION` is the version of release plugins https://github.com/containernetworking/plugins/releases
`CNI_VERSION` is the spec version

Comment (adisky) [reply hack/local-up-cluster.sh]: removed `CNI_VERSION`

**Comment by aojea** [comment]
LGTM but ... can we try to get the job fixed in this PR adding new commits?

now it fails because it has the wrong socket endpoint for the runtime

> I0222 04:26:41.859392  141426 remote_runtime.go:71] "Connecting to runtime service" endpoint="unix:///run/containerd/containerd.sock"
W0222 04:26:41.859654  141426 clientconn.go:1331] [core] grpc: addrConn.createTransport failed to connect to {/run/containerd/containerd.sock /run/containerd/containerd.sock <nil> 0 <nil>}. Err: connection error: desc = "transport: Error while dialing dial unix /run/containerd/containerd.sock: connect: no such file or directory". Reconnecting...
Error: failed to run Kubelet: unable to determine runtime API version: rpc error: code = Unavailable desc = connection error: desc = "transport: Error while dialing dial unix /run/containerd/containerd.sock: connect: no such file or directory"
Usage:

**Comment by aojea** [comment]
ahh, the problem is not kubelet, is containerd, right?

/lgtm

**Comment by adisky** [comment]
@aoeja have you seen this PR https://github.com/kubernetes/test-infra/pull/25352
the job would get fixed when we merge both

**Pull request merged**